src/signature.sh

```diff
@@ -0,0 +1,402 @@
+    if [[ "$?" -gt 0 ]]; then
+      return 1
+    fi
+    trailer+="$formated_output"
+  else
+    signature="$(str_strip "$signature")"
+    if [[ "$keyword" != 'FIXES' ]] && ! is_valid_signature "$signature"; then
+      return 22 # EINVAL
+    fi
+    trailer+="$signature"
+  fi
+
+  all_trailers+=("$trailer")
+}
+
+# This function receives a commit SHA and verifies if it's
+# a valid commit reference. If it is, then it outputs the
+# appropriate formatted message. Else it returns an error code.
+#
+# @sha Holds either a commit hash or pointer
+#
+# Return:
+# In case of successful return 0 and prints the formatted message,
+# otherwise, return 22.
+function format_fixes_message()
+{
+  local sha="$1"
+  local formatted_message
+
+  # Check if given value is a valid commit reference
+  if [[ $(git cat-file -t "$sha" 2> /dev/null) != 'commit' ]]; then
+    return 22 # EINVAL
+  fi
+
+  # The 'Fixes:' trailer line must follow a format defined by Linux Kernel developers.
+  # Fixes: e21d2170f366 ("video: remove unnecessary platform_set_drvdata()")
+  formatted_message=$(git log -1 "$sha" --oneline --abbrev-commit --abbrev=12 \
+    --format="%h (\\\"%s\\\")")
+
+  printf '%s' "$formatted_message"
+}
+
+# It gets user.name and user.email from git's configuration.
+# If either the name or email are not configured then this
+# function will return an error code. Otherwise it will output
+# the formatted name and email.
+#
+# Return:
+# In case of successful return 0 and prints a properly formatted
+# output, otherwise, return 1.
+function format_name_email_from_user()
+{
+  local user_name
+  local user_email
+
+  user_name="$(git config user.name)"
+  user_email="$(git config user.email)"
+
+  # If user doesn't have either a name or email configured with
+  # git then they must provide an argument
+  if [[ -z "$user_name" || -z "$user_email" ]]; then
+    return 1
+  fi
+
+  printf '%s' "${user_name} <${user_email}>"
+}
+
+# This function writes all the trailer lines stored in @all_trailers
+# into either a patch file or commit. It prints a warning if the
+# given argument is neither a valid patch path nor a commit reference
+# and then ignores it.
+#
+# @patch_or_sha Holds either a patch path or commit SHA
+# @flag Used to specify how 'cmd_manager' will be executed
+function write_all_trailers()
+{
+  local patch_or_sha="$1"
+  local flag="$2"
+  local cmd
+
+  for trailer in "${all_trailers[@]}"; do
+    # Check if given argument is either a patch or valid commit reference,
+    # then build the correct command.
+    if [[ "$(git cat-file -t "$patch_or_sha" 2> /dev/null)" == 'commit' ]]; then
+      cmd="git commit --quiet --amend --no-edit --trailer \"${trailer}\""
+      # Only call 'git rebase' if user is trying to write multiple commits
+      if [[ "$(git rev-parse "$patch_or_sha")" != "$(git rev-parse HEAD)" ]]; then
+        cmd="git rebase ${patch_or_sha} --exec '${cmd}' 2> /dev/null"
```

Comment:
Instead of using /dev/null, can we just use `--quiet`.

Reply:
I've tried this before, but the `git rebase` command prints multiple lines like:
```
Executing: git commit --quiet --amend --no-edit --trailer "Tag: Joe Doe <mail>"
Executing: git commit --quiet --amend --no-edit --trailer "Tag: Bob Doe <mail>"
Executing: git commit --quiet --amend --no-edit --trailer "Tag: Michael Doe <mail>"
```
Even when we use `--quiet`, but redirecting `2` to `/dev/null` solves it. I don't know why it behaves like this to be honest and I didn't take the time to investigate it properly. I'll be coming back and taking a look at this in the future, since this is something that has been bothering me as well since I wrote this code.

Reply:
This can be tested running the following command inside a git project:
```bash
git rebase HEAD~3 --quiet --exec "printf 'blabla' > /dev/null" 
```
You will see that even with `--quiet`, the command still outputs the following:
```text
Executing: printf 'blabla' > /dev/null
Executing: printf 'blabla' > /dev/null
Executing: printf 'blabla' > /dev/null
```

Reply:
Interesting. In that case, we can live with that. Could you add a tiny comment about that to prevent someone from changing it in the future?